Content.Shared/Xenoarchaeology/Artifact/SharedXenoArtifactSystem.Graph.cs

```diff
@@ -0,0 +1,451 @@
+using System.Diagnostics.CodeAnalysis;
+using Content.Shared.Xenoarchaeology.Artifact.Components;
+using Robust.Shared.Prototypes;
+using Robust.Shared.Utility;
+
+namespace Content.Shared.Xenoarchaeology.Artifact;
+
+/// <summary>
+/// User-friendly API for viewing and modifying the complex graph relationship in XenoArtifacts
+/// </summary>
+public abstract partial class SharedXenoArtifactSystem
+{
+    /// <summary>
+    /// Gets the index corresponding to a given node, throwing if the node is not present.
+    /// </summary>
+    public int GetIndex(Entity<XenoArtifactComponent> ent, EntityUid node)
+    {
+        if (TryGetIndex((ent, ent), node, out var index))
+        {
+            return index.Value;
+        }
+
+        throw new ArgumentException($"node {ToPrettyString(node)} is not present in {ToPrettyString(ent)}");
+    }
+
+    public bool TryGetIndex(Entity<XenoArtifactComponent?> ent, EntityUid node, [NotNullWhen(true)] out int? index)
+    {
+        index = null;
+        if (!Resolve(ent, ref ent.Comp))
+            return false;
+
+        for (var i = 0; i < ent.Comp.NodeVertices.Length; i++)
+        {
+            if (!TryGetNode(ent, i, out var iNode))
+                continue;
+
+            if (node != iNode.Value.Owner)
+                continue;
+
+            index = i;
+            return true;
+        }
+
+        return false;
+    }
+
+    public Entity<XenoArtifactNodeComponent> GetNode(Entity<XenoArtifactComponent> ent, int index)
+    {
+        if (ent.Comp.NodeVertices[index] is { } netUid && GetEntity(netUid) is var uid)
+            return (uid, XenoArtifactNode(uid));
+
+        throw new ArgumentException($"index {index} does not correspond to an existing node in {ToPrettyString(ent)}");
```

Comment:
Seeing a throw here without much in the way of handling above it. A tad spooky.

Reply:
well, i honestly am not sure why its not try-get pattern, i was considering this a problem - but during testing i was REALLY not excited of scenario when some code was adressing invalid index and game would just go along with it and say 'okay, i cannot do that, here is a log that you never going to see, bye!'

I obviously can remake it into try-get, with loads of additional code that will be enforced to make additional checks... i am just not sure i really want that...